spec/app/ics-721-nft-transfer/README.md

```diff
@@ -0,0 +1,386 @@
+      abortTransactionUnless(escrowAccount === nft.getOwner(data.classId.slice(len(prefix)), data.tokenIds[i]))
+      // unescrow token to receiver
+      err = nft.Transfer(data.classId.slice(len(prefix)), data.tokenIds[i], data.receiver)
+      if (err !== nil) {
+        ack = NonFungibleTokenPacketAcknowledgement{false, "transfer nft(" + data.classId + ", " + data.tokenIds[i] + ") failed"}
+        break
+      }
+    } else {
+      prefix = "{packet.destPort}/{packet.destChannel}/"
+      prefixedClassId = prefix + data.classId
+      // sender was source, mint voucher to receiver
+      err = nft.Mint(prefixedClassId, data.classUri, data.tokenIds[i], data.tokenUris[i], data.receiver)
+      if (err !== nil) {
+        ack = NonFungibleTokenPacketAcknowledgement{false, "mint nft(" + data.classId + ", " + data.tokenIds[i] + ") failed"}
+        break
+      }
+    }
+  }
+  return ack
+}
+```
+
+`onAcknowledgePacket` is called by the routing module when a packet sent by this module has been acknowledged.
+
+```typescript
+function onAcknowledgePacket(
+  packet: Packet,
+  acknowledgement: bytes) {
+  // if the transfer failed, refund the tokens
+  if (!ack.success)
+    refundToken(packet)
+}
+```
+
+`onTimeoutPacket` is called by the routing module when a packet sent by this module has timed-out (such that it will not be received on the destination chain).
+
+```typescript
+function onTimeoutPacket(packet: Packet) {
+  // the packet timed-out, so refund the tokens
+  refundToken(packet)
+}
+```
+
+`refundToken` is called by both `onAcknowledgePacket`, on failure, and `onTimeoutPacket`, to refund escrowed token to the original sender.
+
+```typescript
+function refundToken(packet: Packet) {
+  NonFungibleTokenPacketData data = packet.data
+  prefix = "{packet.sourcePort}/{packet.sourceChannel}/"
+  for (let tokenId in data.tokenIds) {
+    // we are the source if the classId is not prefixed
+    source = data.classId.slice(0, len(prefix)) !== prefix
+    if source {
+      // sender was source chain, unescrow tokens back to sender
+      escrowAccount = channelEscrowAddresses[packet.srcChannel]
+      // assert that escrow account is token owner
+      abortTransactionUnless(escrowAccount === nft.getOwner(data.classId, tokenId))
+      nft.Transfer(data.classId, tokenId, data.sender)
+    } else {
+      // receiver was source chain, mint voucher back to sender
+      bank.Mint(data.classId, tokenId, data.sender)
+    }
+  }
+}
+```
+
+```typescript
+function onTimeoutPacketClose(packet: Packet) {
+  // can't happen, only unordered channels allowed
+}
+```
+
+#### Reasoning
+
+##### Correctness
+
+This implementation preserves both uniqueness & supply.
+
+Uniqueness: If tokens have been sent to the counterparty chain, they can be redeemed back in the same `classId` & `tokenId` on the source chain.
+
+Supply: Redefine supply as unlocked tokens. All send-recv pairs for any given token class sum to net zero. Source chain can change supply.
```

Comment:
this is probably not quite the right property for NFTs

Reply:
"Supply: ..." removed.